Discord Login

berry-13 opened the pull request:
﻿Please include a summary of the changes and the related issue. Please also include relevant motivation and context. List any dependencies that are required for this change.

Added Discord Login and fix some Github login code formatting. I don't understand why **Discord doesn't give permission to access my profile name and picture**.

## Type of change

Please delete options that are not relevant.

- [x] New feature (non-breaking change which adds functionality)
- [x] Documentation update 

## How Has This Been Tested?

Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration:
##

I tested it on docker desktop and with a fresh install on ubuntu

### **Test Configuration**:
##


## Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [x] Any dependent changes have been merged and published in downstream modules

berry-13 committed:
Add files via upload

berry-13 committed:
Create linode-setup.md

berry-13 committed:
Create cloudflare-setup.md

berry-13 committed:
Update cloudflare-setup.md

berry-13 committed:
Delete 4-linode.png

berry-13 committed:
Delete 3-linode.png

berry-13 committed:
Add files via upload

berry-13 committed:
Add files via upload

berry-13 committed:
Update cloudflare-setup.md

berry-13 committed:
Update linode-setup.md

berry-13 committed:
Rename cloudflare-setup.md to cloudflare.md

berry-13 committed:
Rename linode-setup.md to linode.md

berry-13 committed:
Update mkdocs.yml

berry-13 committed:
Update cloudflare.md

berry-13 committed:
Update linode.md

berry-13 committed:
Update README.md

berry-13 committed:
Update README.md

berry-13 committed:
Update linode.md

sentence in Italian

berry-13 committed:
Merge branch 'danny-avila:main' into main

berry-13 committed:
Merge branch 'danny-avila:main' into main

berry-13 committed:
v1

The frontend has been completed, along with the .env variables.

However, there is an issue of infinite loading thereafter.

berry-13 committed:
Fix email and remove deprecated GitHub passport

berry-13 committed:
Merge branch 'danny-avila:main' into main

berry-13 committed:
Update user_auth_system.md

add How to Set Up a Github Authentication

berry-13 committed:
Update .env.example

Improved the comment above the GitHub client ID and secret.

berry-13 committed:
Update user_auth_system.md

berry-13 committed:
Update package.json

berry-13 committed:
Remove unnecessary passport GitHub package

berry-13 committed:
Merge branch 'main' of http://github.com/Berry-13/librechat

berry-13 committed:
Merge branch 'danny-avila:main' into main

berry-13 committed:
fixed conflicts

 fixed conflicts between Berry-13:main and danny-avila:main

in api/server/index.js 45:54

berry-13 committed:
Delete e -i HEAD~2

berry-13 committed:
Merge branch 'main' into main

berry-13 committed:
(WIP) Discord Login

berry-13 committed:
Merge branch 'main' of http://github.com/berry-13/LibreChat

berry-13 committed:
Merge branch 'danny-avila:main' into main

berry-13 committed:
Merge branch 'danny-avila:main' into main

berry-13 committed:
Fix duplicate githubLoginEnabled

berry-13 committed:
.env.example restore

berry-13 committed:
Update user_auth_system.md

Discord Login

Berry-13 commented:
Hi Danny, since from what I have experienced, it appears impossible to have a name and a profile image with Discord Application , we could use the email address as the name? (is it the same on chat.openai.com as well)

danny-avila commented:
> Hi Danny, since from what I have experienced, it appears impossible to have a name and a profile image with Discord Application , we could use the email address as the name? (is it the same on chat.openai.com as well)

To clarify, are you saying this is the default behavior on chat.openai.com?

I'm actually pretty experienced with discord development and I can tell you that they are very restrictive with their scoping. I will take a look at your code but this looks to be part of the solution, you may have looked at it already: https://stackoverflow.com/questions/68379109/javascript-discord-get-profile-picture

danny-avila commented on api/strategies/discordStrategy.js:
```diff
@@ -0,0 +1,38 @@
+        username: profile.username,   // username is the username (the old tag discord#1111) of the user on discord
+        email,
+        name: profile.global_name,  // global_name is the name of the user on discord
+        avatar: `https://cdn.discordapp.com/avatars/${profile.id}/${profile.avatar}.png`
```
you should handle the edge case where a user has no avatar set, and this URL should be assigned outside of this object

danny-avila replied:
I haven't tested this but according to GPT, this may help with default avatars

```js
      let avatarURL;
      if (profile.avatar) {
        const format = profile.avatar.startsWith('a_') ? 'gif' : 'png';
        avatarURL = `https://cdn.discordapp.com/avatars/${profile.id}/${profile.avatar}.${format}`;
      } else {
        const defaultAvatarNum = Number(profile.discriminator) % 5;
        avatarURL = `https://cdn.discordapp.com/embed/avatars/${defaultAvatarNum}.png`;
      }

      console.debug('Discord profile', profile);
      console.debug('Discord avatarURL', avatarURL);
```

**From GPT:**
When a user doesn't have a custom avatar, Discord provides a default one. There are five default avatars, and the one that's used for a particular user is determined by the last digit of their discriminator (the four-digit number that comes after their username), modulo 5.

Here's the relevant part of the Discord Developer Documentation that talks about this:

> "If the user has not set an avatar, the `avatar` field will be `null` and you can generate a default URL by concatenating the user's discriminator modulo 5 to the URL `https://cdn.discordapp.com/embed/avatars/.png`."

You can find this in the [User Object section](https://discord.com/developers/docs/resources/user#user-object) of the Discord Developer Documentation.

So, in the code snippet:

```javascript
const defaultAvatarNum = Number(profile.discriminator) % 5;
avatarURL = `https://cdn.discordapp.com/embed/avatars/${defaultAvatarNum}.png`;
```

`defaultAvatarNum` is the user's discriminator modulo 5, which determines which of the five default avatars to use. `avatarURL` is then the URL of this default avatar.

danny-avila commented on docs/install/user_auth_system.md:
```diff
@@ -80,6 +80,18 @@ OPENID_CALLBACK_URL=/oauth/openid/callback
+
+1. Go to [Discord Developer Portal](https://discord.com/developers)
+2. Create a new Application and give it a name
+4. In the OAuth2 general settings add a redirect URL and set it as "[Your DOMAIN_CLIENT](https://github.com/danny-avila/LibreChat/blob/main/.env.example#L219)/oauth/discord/callback" (example: http://localhost:3080/oauth/discord/callback)
```
as @fuegovic mentioned on discord, .env.example should be updated, too. What i said above is in addition to that, as it would be helpful to have them in the instructions

danny-avila commented on client/src/components/Auth/Registration.tsx:
```diff
@@ -347,33 +347,59 @@ function Registration() {
+                href={`${startupConfig.serverDomain}/oauth/github`}
+                className="justify-left flex w-full items-center space-x-3 rounded-md border border-gray-300 px-5 py-3 hover:bg-gray-50 focus:ring-2 focus:ring-violet-600 focus:ring-offset-1"
+              >
+                <svg
```
It would be extremely helpful for me if we move the SVG logic (each social icon) into its own component in `client\src\components\svg` and exported from `client\src\components\svg\index.ts`

this would clean up a lot of the logic

danny-avila replied:
you would then be able to import them like this
`import { GitHubIcon, DiscordIcon, /* ...rest */ } from '~/components'`

berry-13 committed:
Fix and new Feature

1. Added Discord login to .env.example.
2. Created Google, Github, and Discord icons in client\src\components\svg.
3. Added the social login option in the .env file; it fixes the ---or---. Check Discord for more information.

berry-13 committed:
fix Login.tsx and Registration.tsx

berry-13 committed:
Update user_auth_system.md

berry-13 committed:
Update .env.example

berry-13 committed:
Added OpenID Icon

berry-13 committed:
quick discord icon fix

berry-13 committed:
discord strategy fix

danny-avila commented on api/strategies/discordStrategy.js:
```diff
@@ -0,0 +1,61 @@
+        return cb(null, existingUser);
+      }
+
+      // Controllo se esiste un utente con la stessa email ma diverso discordId
```
small nitpick, but removing the comments here would be good

danny-avila commented:
Looks good to me, just remove those comments and I can merge

danny-avila commented:
there are now 2 minor merge conflicts from a PR i just approved but they should be easy to fix

berry-13 committed:
Merge branch 'main' into main

Berry-13 commented:
Fixed! If you can, give me another hour because I'm testing one last thing with the login. Thanks.

berry-13 committed:
Merge branch 'main' of http://github.com/Berry-13/LibreChat

danny-avila commented:
> Fixed! If you can, give me another hour because I'm testing one last thing with the login. Thanks.

sounds good, thanks!

berry-13 committed:
remove comment
